core/vm/ipgraph_with_kind.go

```diff
@@ -0,0 +1,365 @@
+		return c.ipGraph.RequiredGas(input)
+	case bytes.Equal(selector, getAncestorIpsSelector):
+		return c.ipGraph.RequiredGas(input)
+	case bytes.Equal(selector, getAncestorIpsCountSelector):
+		return c.ipGraph.RequiredGas(input)
+	case bytes.Equal(selector, hasAncestorIpsSelector):
+		return c.ipGraph.RequiredGas(input)
+	case bytes.Equal(selector, setRoyaltyWithKindSelector):
+		return ipGraphWriteGas
+	case bytes.Equal(selector, getRoyaltyWithKindSelector):
+		royaltyPolicyKind := new(big.Int).SetBytes(getData(input, 64, 32))
+		if royaltyPolicyKind.Cmp(royaltyPolicyKindLAP) == 0 {
+			return ipGraphReadGas * (averageAncestorIpCount * 3)
+		} else if royaltyPolicyKind.Cmp(royaltyPolicyKindLRP) == 0 {
+			return ipGraphReadGas * (averageAncestorIpCount*2 + 2)
+		} else {
+			return intrinsicGas
+		}
+	case bytes.Equal(selector, getRoyaltyStackWithKindSelector):
+		royaltyPolicyKind := new(big.Int).SetBytes(getData(input, 64, 32))
+		if royaltyPolicyKind.Cmp(royaltyPolicyKindLAP) == 0 {
+			return ipGraphReadGas * (averageParentIpCount + 1)
+		} else if royaltyPolicyKind.Cmp(royaltyPolicyKindLRP) == 0 {
+			return ipGraphReadGas * (averageAncestorIpCount * 2)
+		} else {
+			return intrinsicGas
+		}
+	default:
+		return intrinsicGas
+	}
+}
+
+func (c *ipGraphWithPolicyKind) Run(evm *EVM, input []byte) ([]byte, error) {
+	ipGraphAddress := common.HexToAddress("0x000000000000000000000000000000000000001B")
+	log.Info("ipGraph.Run", "input", input)
+
+	if len(input) < 4 {
+		return nil, fmt.Errorf("input too short")
+	}
+
+	selector := input[:4]
+	args := input[4:]
+
+	switch {
+	case bytes.Equal(selector, addParentIpSelector):
+		return c.ipGraph.addParentIp(args, evm, ipGraphAddress)
+	case bytes.Equal(selector, hasParentIpSelector):
+		return c.ipGraph.hasParentIp(args, evm, ipGraphAddress)
+	case bytes.Equal(selector, getParentIpsSelector):
+		return c.ipGraph.getParentIps(args, evm, ipGraphAddress)
+	case bytes.Equal(selector, getParentIpsCountSelector):
+		return c.ipGraph.getParentIpsCount(args, evm, ipGraphAddress)
+	case bytes.Equal(selector, getAncestorIpsSelector):
+		return c.ipGraph.getAncestorIps(args, evm, ipGraphAddress)
+	case bytes.Equal(selector, getAncestorIpsCountSelector):
+		return c.ipGraph.getAncestorIpsCount(args, evm, ipGraphAddress)
+	case bytes.Equal(selector, hasAncestorIpsSelector):
+		return c.ipGraph.hasAncestorIp(args, evm, ipGraphAddress)
+	case bytes.Equal(selector, setRoyaltyWithKindSelector):
+		return c.setRoyaltyWithKind(args, evm, ipGraphAddress)
+	case bytes.Equal(selector, getRoyaltyWithKindSelector):
+		return c.getRoyaltyWithKind(args, evm, ipGraphAddress)
+	case bytes.Equal(selector, getRoyaltyStackWithKindSelector):
+		return c.getRoyaltyStackWithKind(args, evm, ipGraphAddress)
+	default:
+		return nil, fmt.Errorf("unknown selector")
+	}
+}
+
+func (c *ipGraphWithPolicyKind) setRoyaltyWithKind(input []byte, evm *EVM, ipGraphAddress common.Address) ([]byte, error) {
```

Comment:
Would you kindly add some comments about what "kind" means? It will be helpful for reviewers and auditors

Reply:
added comments.